### DISCUSSION TITLE
Dynamically inject environment variable into Vercel git integration builds

### DISCUSSION THREAD

**Discussion by ryancausey** [vercel/vercel#8801] [Help] [answered]
There seems to be a few related questions, but I wasn't quite able to get what I was looking for out of them. If this is already answered in one of those other discussions and I missed it, I apologize:

https://github.com/vercel/vercel/discussions/4853
https://github.com/vercel/vercel/discussions/6534
https://github.com/vercel/vercel/discussions/8543

I'm trying to see if there's a middle ground between using the GitLab integration and going full custom on GitLab CI/CD for my use case:

When someone opens a GitLab merge request (MR), I have the current CI/CD pipeline spin up a copy of the back-end API with a unique URL based on the MR ID. I know this URL at the start of the pipeline (on push). I need a way to dynamically inject this URL into the Vercel build as an environment variable on a per-branch basis.

It seems like the right approach would be using [vercel env](https://vercel.com/docs/cli/env). However, since the Vercel build would kick off on push in parallel with the GitLab CI/CD pipeline starting, it seems like there would be a race condition between the GitLab CI/CD job running the `vercel env add [name] [environment] [gitbranch]` command and the Vercel build starting (and failing if the env var isn't there).

Is there some other method to inject this environment variable that I'm not aware of?

**Comment by amyegan** [comment, marked as answer]
Thanks for reaching out @ryancausey! I don't think that `vercel env` will work if you use it after starting the [build](https://vercel.com/docs/cli/build) since the app would already be built with an older (or non-existent) value. But if you can set the variable before running `vercel pull` and `vercel build`, then it should work.

**Reply by ryancausey** [reply]
@amyegan with the Git integration, is there a way to delay the build/deploy/other stuff until I take the action to set up the environment variable?

**Reply by amyegan** [reply]
You could do something fancy with [Ignored Build Step](https://vercel.com/guides/how-do-i-use-the-ignored-build-step-field-on-vercel). I'm sure there are tons of ways that other people have created to accomplish similar goals, but here are a couple of options I could think of:

1) You could skip all automatic deployments except Production like in [this env var example](https://vercel.com/guides/how-do-i-use-the-ignored-build-step-field-on-vercel#with-environment-variables). Then you would still need to switch to the [pipeline](https://vercel.com/guides/how-can-i-use-gitlab-pipelines-with-vercel) for preview deployments, but could still have the added features of the git integration for production deployments.

2) You could do something like https://github.com/vercel/community/discussions/60#discussioncomment-3643433, but instead of using the name tag to skip a build you could use it as the way to trigger a preview build. You would skip the initial MR build, then create an extra commit with something like `[deploy]` in the message from your current pipeline once the API URL is available and set as an environment variable for the branch, then let the newly created commit trigger the preview deployment. 😅 Feels a bit like a Rube Goldberg machine, but it should work if you don't mind the extra commit clutter on your MR.

**Reply by ryancausey** [reply]
@amyegan yes, it does seem somewhat rube-goldbergy, but I'm essentially trying to see if there's a way to avoid having to opt out of using the Vercel platform's build system. My understanding is that dropping into using `vercel build` in GitLab CI/CD would mean building locally on GitLab runners instead of using the Vercel platform to do the build. I don't really want to have to opt out of this feature of the Vercel platform because of needing to inject a per-branch dynamic environment variable. It would seem silly to not leverage what appears to be a highly optimized build system for next.js projects.

I'll look into all these options and see if I can't find some solution.

**Reply by amyegan** [reply]
I know what you mean. The problem is getting the environment variable set before the build starts. You could use the legacy [env](https://vercel.com/docs/project-configuration#legacy/env) setting in a `vercel.json` config, but that would still require an extra commit after the backend URL is known. You could take a look at [Deploy Hooks](https://vercel.com/docs/concepts/git/deploy-hooks#) and see if that would work for you. But that has a per branch setup designed more for headless CMS updates and cron jobs. Or you could try [creating a deployment with the REST API](https://vercel.com/docs/rest-api#endpoints/deployments/create-a-new-deployment).

On the other side of things, maybe there is a way to alias your backend service to a predictable staging URL. That way you wouldn't need to wait until after the backend service is deployed to set the env var.
